### ISSUE TITLE
Beta 4.48.15 crashes on launch

### ISSUE THREAD

**Issue by brs17** [signalapp/Signal-Android#9085]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
After updating to 4.48.15 this afternoon, Signal now crashes just about immediately after attempting to launch it.

### Steps to reproduce
- Select Signal App

**Actual result:** 
After selecting Signal, the screen turns white as though the program is launching, after a second or so, the application crashes and I am presented with a `Signal has stopped` pop-up notification.

**Expected result:**
Signal should launch

### Device info
**Device:** LG Nexus 5x
**Android version:** 8.1.0
**Signal version:** 4.48.15

### Link to debug log
Below are the logs that seem directly related to the crash as captured with `adb logcat`
[signalcrash.log](https://github.com/signalapp/Signal-Android/files/3705516/signalcrash.log)

**Comment by greyson-signal** [comment]
In the middle of getting a fix together right now.

- Do you know what version of Signal you were running before upgrading to 4.48.15?
- Can you tell me if you had an funny-looking conversations before this update happened?

**Comment by brs17** [comment]
> In the middle of getting a fix together right now.
> 
>     * Do you know what version of Signal you were running before upgrading to 4.48.15?
> 
>     * Can you tell me if you had an funny-looking conversations before this update happened?

1. I am pretty sure I updated Signal yesterday, so my guess would be 4..48.14 although I am not certain.
2. The content of the conversations might have been "funny", but I can't say I noticed anything out of place about the conversations. Although to be fair I have not played around much with the new 4.48 features.
3. I did just discover that I can still receive Signal notifications to my phone and can respond to messages that way.

**Comment by alex7899** [comment]
Same question from me!

1. Before 4.48.14
2. No

**Comment by greyson-signal** [comment]
I just published 4.48.16. Usually takes a little bit for the Play Store to actually let people install it, but it shouldn't be too long. After updating, let me know if you're still experiencing this issue. Thank you!

**Comment by greyson-signal** [comment]
@brs17 @alex7899 

After updating, I'm assuming you're going to have some conversation in your conversation list with the label "Unknown". If you do, let me know if there's messages in that conversation, whether or not you recognize them as coming from a specific person, what the history might have been there, etc. Thanks!

**Comment by RiseT** [comment]
Same issue.

1. Prior version 4.48.14
2. No funny looking messages
3. Updating to 4.48.16 fixed the issue
4. No conversations labeled as "Unknown" or anything else unusual

**Comment by johanw666** [comment]
I see one chat with "Unknown". It contains some automatically received sms messages from 3 different senders from 2014.

**Comment by brs17** [comment]
> @brs17 @alex7899
> 
> After updating, I'm assuming you're going to have some conversation in your conversation list with the label "Unknown". If you do, let me know if there's messages in that conversation, whether or not you recognize them as coming from a specific person, what the history might have been there, etc. Thanks!

@greyson-signal, you published the patch after I turned in for the night, and I just saw the update in the Google Play store this morning, updated, and your patch seems to have fixed the issue!

I do not see a conversation labeled "Unknown", all of my messages seem to be in the right spot from the right people. For what it is worth, I only currently have one conversation that is not archived. Conversations that are currently archived do not seem to have an "Unknown" label issue either.

Thank you!

**Comment by u32i64** [comment]
For me, no `Unknown` chats in conversation list, archived conversations, or contacts (on the "new message" screen). SMS integration not enabled. Previous version to 4.48.15 that I had installed was 4.48.14. Conversations appear to be normal.

**Comment by alex7899** [comment]
@greyson-signal 

For me, no Unknown chats in conversation list, after updating to 4.48.16, Now signal is working.  Thanks for your quick reply!

**Comment by greyson-signal** [comment]
Thank you all for the information! I expected something like @johanw666's situation, where some weirdly-formatted SMS/MMS created some strange threads that I didn't account for in a 4.48.15 commit. Strange indeed that many of you that experienced the crash don't have any "unknown" chats. I shall think more on this, but I'm happy that the crash is at least gone :)

**Comment by RiseT** [comment]
I do have a couple with - like @johanw666 mentioned - a name as the sender (one of them is "github" btw), no "Unknown" labels, though (where is the "Unknown" label supposed to get displayed?).

**Comment by pgerber** [comment]
I've had another look at the error message. So, this is somehow related to a missing recipient (`MissingRecipientError`). The most recent message in one of the conversations is a contact I sent. Could it be that the error occurs because Signal tries to fetch information about that contact when showing the preview for that message? That could explain why Signal thinks a recipient is missing; that user is not on Signal.

**Comment by tsimonq2** [comment]
Thanks for the work on this, I can confirm that it's now fixed.

**Comment by greyson-signal** [comment]
The crash was happening because the "recipient ID" (a locally-generated ID for someone you've interacted with) associated with a thread in the conversation list was -1, which indicates an "unknown" recipient. I naively thought those scenarios shouldn't happen anymore, so I took out the handling for unknown recipients, which caused the crashes ya'll experienced. I added back and improved the handling to fix the crash.

I have some theories as to how the threads had unknown recipients, and I've prevented future threads from being created with unknown recipients. My assumption is that we had some old badly-parsed MMS message or something, or possibly created a bad thread due to a race condition that was present in an earlier release of 4.48. I think we're safe now though, so thank you all for the information!

**Comment by johanw666** [comment]
> I do have a couple with - like @johanw666 mentioned - a name as the sender (one of them is "github" btw), no "Unknown" labels, though (where is the "Unknown" label supposed to get displayed?).

I also have senders with non-numbers in the address field that bare nicely grouped in their own conversation, like all other messages I got from that 2FA organisation when I still used Signal for sms (up to TextSecure 2.6.4). Those messages have their own thread with the sender's name as title, so nun-numeric senders can not be the whole story.

I see the same messages in (I imported them during the split after TS 2.6.4) Silence, there the threading seems OK.
